### DISCUSSION TITLE
Unable to boot Jetson Xavier NX Devkit from SD card

### DISCUSSION THREAD

**Discussion by NavSriv** [OE4T/meta-tegra-community#106] [Q&A] [answered]
I was able to boot jetson xavier NX devkit from SD card by using custom image created from yocto. I followed the instruction available in  [link](https://github.com/OE4T/meta-tegra/wiki/initrd-flashing-support) and was able to boot from SSD also by using custom image. **Now I want to boot from SD card again, but board is flashing by default from SSD**. 

I re-flashed SPI from SD card image package by using "./doflash.sh" assuming it might change configurations to point rootfs to SD card but that didn't work. I realised there was a service mounting rootfs to SSD and I disabled it but still no luck. Am I missing something here?

**Comment by madisongh** [comment, marked as answer]
If you mean that the device is still *booting* from the SSD (or NVMe drive), there are a couple of reasons that could happen.

One is the UEFI bootloader's boot order, which by default puts SSDs/NVME drives before SD/eMMC. You can modify behavior that it in the UEFI boot manager settings.

The other is the boot script we provide in the minimal initramfs, which by default just searches for a partition called `APP` to boot from. If you have multiple drives with an `APP` partition (which you would, if you flashed twice to two different storage devices, without erasing one of the drives), the selection of the storage device is not deterministic. You can modify this behavior adding `KERNEL_ARGS:append = " root=/dev/XXXXXXX"` (probably `/dev/mmcblk0p1` in your case) to your build configuration, to force the selection of the partition on the SDcard.

The other thing you can do is to erase the SSD/NVMe drive, or rename the `APP` partition on that drive to something else, so the boot script doesn't try to use it.

You could also provide a replacement for the the boot script (e.g., with a bbappend for `tegra-minimal-init` recipe) that implements whatever logic you need for your particular setup.

**Reply by NavSriv** [reply]
@madisongh 
> You can modify this behavior adding KERNEL_ARGS:append = " root=/dev/XXXXXXX" (probably /dev/mmcblk0p1 in your case) to your build configuration, to force the selection of the partition on the SDcard.

I made `KERNEL_ARGS:append = " root=/dev/mmcblk0p1"` in my local.conf and following appeared on dmesg

`[    0.000000] Kernel command line: console=ttyTCU0,115200 console=tty0 fbcon=map:0 root=/dev/mmcblk0p1 console=ttyTCU0,115200 console=tty0 fbcon=map:0 root=/dev/mmcblk0p1 video=tegrafb earlycon=tegra_comb_uart,mmio32,0x0c168000 gpt rootfs.slot_suffix= tegra_fbmem=0x800000@0xa06ac000 lut_mem=0x2008@0xa06a6000 usbcore.old_scheme_first=1 tegraid=19.1.2.0.0 maxcpus=6 boot.slot_suffix= boot.ratchetvalues=0.4.2 vpr_resize sdhci_tegra.en_boot_part_access=1  
`
which looks fine.

Can you please suggest what I can do in order to load rootfs also from SD card? Right now nvme partition is mounted on /.
Thank you

**Reply by dwalkes** [reply]
I'm guessing this is due to the logic [here](https://github.com/OE4T/meta-tegra/blob/971f01409ba9839c192d5abe7ffe93e355100266/recipes-core/initrdscripts/tegra-minimal-init/platform-preboot.sh#L4-L34) which looks at the `/sys/firmware/efi/efivars/BootChainOsCurrent-781e084c-a330-417c-b678-38e696380cb9` var to determine the boot slot when booting with redundant partitions and `boot.slot_suffix=` specified, then overrides the rootfs based on partition labels, which may end up with the partition on NVMe if you have multiple labeled `APP_slot_suffix`.

> The other thing you can do is to erase the SSD/NVMe drive, or rename the APP partition on that drive to something else, so the boot script doesn't try to use it.

This is still probably your best bet, just remove the partition or rename it on the NVMe drive.

**Reply by NavSriv** [reply]
Formatting the SSD drive helped. I am able to mount rootfs from SD card now. I will look over if I can make changes to tegra-minimal-init in order to change boot.slot-suffix

Thanks @madisongh  and @dwalkes  for the lead.

**Reply by madisongh** [reply]
@dwalkes That we execute that bit of the script all the time, even if A/B isn't enabled, should probably be considered a bug.  I'm not sure what the right fix is... it's probably either letting `root=` override the A/B slotting, or adding a check of the right EFI var to see if A/B is enabled.
